library/std/src/os/windows/io/handle.rs

```diff
@@ -0,0 +1,427 @@
+#[repr(transparent)]
+#[unstable(feature = "io_safety", issue = "87074")]
+pub struct OptionFileHandle {
+    handle: RawHandle,
+}
+
+// The Windows [`HANDLE`] type may be transferred across and shared between
+// thread boundaries (despite containing a `*mut void`, which in general isn't
+// `Send` or `Sync`).
+//
+// [`HANDLE`]: std::os::windows::raw::HANDLE
+unsafe impl Send for OwnedHandle {}
+unsafe impl Send for OptionFileHandle {}
+unsafe impl Send for BorrowedHandle<'_> {}
+unsafe impl Sync for OwnedHandle {}
+unsafe impl Sync for OptionFileHandle {}
+unsafe impl Sync for BorrowedHandle<'_> {}
+
+impl BorrowedHandle<'_> {
+    /// Return a `BorrowedHandle` holding the given raw handle.
+    ///
+    /// # Safety
+    ///
+    /// The resource pointed to by `handle` must be a valid open handle, it
+    /// must remain open for the duration of the returned `BorrowedHandle`, and
+    /// it must not be null.
+    ///
+    /// Note that it *may* have the value `INVALID_HANDLE_VALUE` (-1), which is
+    /// sometimes a valid handle value. See [here] for the full story.
+    ///
+    /// [here]: https://devblogs.microsoft.com/oldnewthing/20040302-00/?p=40443
+    #[inline]
+    #[unstable(feature = "io_safety", issue = "87074")]
+    pub unsafe fn borrow_raw_handle(handle: RawHandle) -> Self {
+        assert!(!handle.is_null());
+        Self { handle: NonNull::new_unchecked(handle), _phantom: PhantomData }
+    }
+}
+
+impl OptionFileHandle {
+    /// Return an empty `OptionFileHandle` with no resource.
+    #[inline]
+    #[unstable(feature = "io_safety", issue = "87074")]
+    pub const fn none() -> Self {
+        Self { handle: c::INVALID_HANDLE_VALUE }
+    }
+}
+
+impl TryFrom<OptionFileHandle> for OwnedHandle {
+    type Error = ();
+
+    #[inline]
+    fn try_from(option: OptionFileHandle) -> Result<Self, ()> {
+        let handle = option.handle;
+        forget(option);
+        if let Some(non_null) = NonNull::new(handle) {
+            if non_null.as_ptr() != c::INVALID_HANDLE_VALUE {
+                Ok(Self { handle: non_null })
+            } else {
+                Err(())
+            }
+        } else {
+            // In theory, we ought to be able to assume that the pointer here
+            // is never null, change `option.handle` to `NonNull`, and obviate
+            // the the panic path here. Unfortunately, Win32 documentation
+            // doesn't explicitly guarantee this anywhere.
+            //
+            // APIs like [`CreateFileW`] itself have `HANDLE` arguments where a
+            // null handle indicates an absent value, which wouldn't work if
+            // null were a valid handle value, so it seems very unlikely that
+            // it could ever return null. But who knows?
+            //
+            // [`CreateFileW`]: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew
+            panic!("An `OptionFileHandle` was null!");
+        }
+    }
+}
```

Comment:
```suggestion
    fn try_from(handle_or_invalid: HandleOrInvalid) -> Result<Self, ()> {
        // In theory, we ought to be able to assume that the pointer here is
        // never null, use `OwnedHandle` rather than `Option<OwnedHandle>`, and
        // obviate the the panic path here.  Unfortunately, Win32 documentation
        // doesn't explicitly guarantee this anywhere.
        //
        // APIs like [`CreateFileW`] itself have `HANDLE` arguments where a
        // null handle indicates an absent value, which wouldn't work if null
        // were a valid handle value, so it seems very unlikely that it could
        // ever return null. But who knows?
        //
        // [`CreateFileW`]: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew
        let owned_handle = handle_or_invalid.0.expect("A `HandleOrInvalid` was null!");
        if owned_handle.0.as_ptr() == c::INVALID_HANDLE_VALUE {
            Err(())
        } else {
            owned_handle
        }
    }
}
```

(Might also be nice to have a function `OwnedHandle::is_invalid`, but I only think that's worth factoring out if we want to make it public, and ideally anyone who might otherwise want to call that should be using `HandleOrInvalid` instead.)